bootstrap/controllers/machine_deployment_controller.go

```diff
@@ -0,0 +1,414 @@
+	}, nil
+}
+
+// getCluster gets the Cluster object for the MachineDeployment.
+func (r *MachineDeploymentReconciler) getCluster(ctx context.Context, md *clusterv1.MachineDeployment) (*clusterv1.Cluster, error) {
+	cluster := &clusterv1.Cluster{}
+	clusterKey := client.ObjectKey{
+		Namespace: md.Namespace,
+		Name:      md.Spec.ClusterName,
+	}
+	if err := r.Get(ctx, clusterKey, cluster); err != nil {
+		return nil, fmt.Errorf("failed to get object: %w", err)
+	}
+
+	return cluster, nil
+}
+
+// getOwnedMachines gets the machines owned by the MachineDeployment.
+func (r *MachineDeploymentReconciler) getOwnedMachines(ctx context.Context, md *clusterv1.MachineDeployment) ([]*clusterv1.Machine, error) {
+	cluster, err := r.getCluster(ctx, md)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster: %w", err)
+	}
+
+	// NOTE(Hue): The machines are not owned by the MachineDeployment directly, but by the MachineSet.
+	var (
+		msList   clusterv1.MachineSetList
+		selector = map[string]string{
+			clusterv1.ClusterNameLabel:           cluster.Name,
+			clusterv1.MachineDeploymentNameLabel: md.Name,
+		}
+	)
+	if err := r.List(ctx, &msList, client.InNamespace(cluster.Namespace), client.MatchingLabels(selector)); err != nil {
+		return nil, fmt.Errorf("failed to get MachineSetList: %w", err)
+	}
+
+	var (
+		ms    clusterv1.MachineSet
+		found bool
+	)
+	// NOTE(Hue): The nosec is due to a false positive: https://stackoverflow.com/questions/62446118/implicit-memory-aliasing-in-for-loop
+	for _, _ms := range msList.Items { // #nosec G601
+		if util.IsOwnedByObject(&_ms, md) {
+			ms = _ms
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		return nil, fmt.Errorf("failed to find MachineSet owned by MachineDeployment %q", md.Name)
+	}
+
+	ownedMachinesCollection, err := r.machineGetter.GetMachinesForCluster(ctx, client.ObjectKeyFromObject(cluster), collections.OwnedMachines(&ms))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get cluster machines: %w", err)
+	}
+
+	ownedMachines := make([]*clusterv1.Machine, len(ownedMachinesCollection))
+	i := 0
+	for _, m := range ownedMachinesCollection {
+		ownedMachines[i] = m
+		i++
+	}
+
+	// NOTE(Hue): Sorting machines by their UID to make sure we have a deterministic order.
+	// This is to (kind of) make sure we upgrade the machines in the same order every time.
+	// Meaning that if in the previous reconciliation we annotated a machine with upgrade-to,
+	// In the next reconciliation we will make sure that upgrade was successful before moving
+	// to the next machine.
+	// This is not the most robust way to do this, but it's good enough.
+	// A better way to do this might be to use some kind of lock (via a secret or something),
+	// similar to control plane init lock.
+	slices.SortStableFunc(ownedMachines, func(m1, m2 *clusterv1.Machine) int {
+		switch {
+		case m1.UID < m2.UID:
+			return -1
+		case m1.UID == m2.UID:
+			return 0
+		default:
+			return 1
+		}
+	})
```

Comment:
Are there any guarantees that UID's increase monotonically? Looks like that's a prerequisite for this sort to work.